infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java

```diff
@@ -72,32 +88,92 @@ public KubernetesNamespaceFactory(
       @Nullable @Named("che.infra.kubernetes.namespace.default") String defaultNamespaceName,
       @Named("che.infra.kubernetes.namespace.allow_user_defined")
           boolean allowUserDefinedNamespaces,
-      KubernetesClientFactory clientFactory)
+      KubernetesClientFactory clientFactory,
+      WorkspaceManager workspaceManager)
       throws ConfigurationException {
     this.namespaceName = namespaceName;
-    this.isPredefined = !isNullOrEmpty(namespaceName) && hasNoPlaceholders(this.namespaceName);
     this.serviceAccountName = serviceAccountName;
     this.clusterRoleName = clusterRoleName;
     this.clientFactory = clientFactory;
     this.defaultNamespaceName = defaultNamespaceName;
     this.allowUserDefinedNamespaces = allowUserDefinedNamespaces;
+    this.workspaceManager = workspaceManager;
+
+    // This will disappear once we support user selection of workspaces...
+    if (allowUserDefinedNamespaces) {
+      LOG.warn(
+          "'che.infra.kubernetes.namespace.allow_user_defined' is not supported yet. It currently has no"
+              + " effect.");
+    }
+
+    // right now allowUserDefinedNamespaces can't be true, but eventually we will implement it.
     if (isNullOrEmpty(defaultNamespaceName) && !allowUserDefinedNamespaces) {
       throw new ConfigurationException(
           "che.infra.kubernetes.namespace.default or "
               + "che.infra.kubernetes.namespace.allow_user_defined must be configured");
     }
   }
 
-  private boolean hasNoPlaceholders(String namespaceName) {
-    return NAMESPACE_NAME_PLACEHOLDERS.keySet().stream().noneMatch(namespaceName::contains);
+  private boolean hasPlaceholders(String namespaceName) {
+    return namespaceName != null
+        && NAMESPACE_NAME_PLACEHOLDERS.keySet().stream().anyMatch(namespaceName::contains);
+  }
+
+  /** True if namespace is potentially created for the workspaces, false otherwise. */
+  protected boolean isCreatingNamespaces(String workspaceId) throws InfrastructureException {
```

Comment:
Well, I personally try to order the methods by visibility, because that hints at their importance in the conjunction with the rest of the codebase. I left the `hasPlaceholders` where it was originally to minize the diff.